### PR TITLE
Replacing Table join by User id, song id

### LINKED ISSUE #14: save ids instead of mapping

### PR DESCRIPTION
**Pull request by krishrustagi**

### CODE REVIEW

File: src/main/java/com/emotunes/emotunes/dao/UserSongMappingDao.java [hunk 1]
```diff
@@ -25,58 +26,64 @@ public class UserSongMappingDao {
     private final UserSongMappingRepository userSongMappingRepository;
 
     public void addSong(String userId, String songId, Emotion emotion) {
-        userSongMappingRepository.save(
-                UserSongMappingMapper.toEntity(userRepository.getReferenceById(userId),
-                        songRepository.getReferenceById(songId), emotion));
+        userSongMappingRepository.save(UserSongMappingMapper.toEntity(userId, songId, emotion));
     }
 
     public List<SongMetadata> getAll(String userId) {
```
Comment (SarthakSavasil) [inline src/main/java/com/emotunes/emotunes/dao/UserSongMappingDao.java]: function Name...

File: src/main/java/com/emotunes/emotunes/dao/UserSongMappingDao.java [hunk 2]
```diff
@@ -25,58 +26,64 @@ public class UserSongMappingDao {
...
-        List<StoredUserSongMapping> songList =
-                userSongMappingRepository.findAllSongsOfUser(userRepository.getReferenceById(userId));
+        List<StoredUserSongMapping> userSongMappingList =
+                userSongMappingRepository.findAllSongsOfUser(userId);
 
         List<SongMetadata> songMetadataList = new ArrayList<>();
```
Comment (SarthakSavasil) [inline src/main/java/com/emotunes/emotunes/dao/UserSongMappingDao.java]: break the function into 2 from here and return the function call directly

File: src/main/java/com/emotunes/emotunes/dao/UserSongMappingDao.java [hunk 3]
```diff
@@ -25,58 +26,64 @@ public class UserSongMappingDao {
...
-        List<StoredUserSongMapping> songList =
-                userSongMappingRepository.findAllSongsWithEmotionOfUser(userRepository.getReferenceById(userId),
-                        emotion.name());
+        List<StoredUserSongMapping> userSongMappingList =
+                userSongMappingRepository.findAllSongsWithEmotionOfUser(userId, emotion.name());
 
         List<SongMetadata> songMetadataList = new ArrayList<>();
 
-        songList.forEach(song -> songMetadataList.add(SongMapper.toSongMetadata(song)));
+        userSongMappingList.forEach(userSongMapping -> {
```
Comment (SarthakSavasil) [inline src/main/java/com/emotunes/emotunes/dao/UserSongMappingDao.java]: Repeated Code.. Use function

File: src/main/java/com/emotunes/emotunes/dao/UserSongMappingDao.java [hunk 4]
```diff
@@ -25,58 +26,64 @@ public class UserSongMappingDao {
...
     public List<SongMetadata> getAllLikedSongs(String userId) {
-        List<StoredUserSongMapping> songList =
-                userSongMappingRepository.findAllLikedSongsOfUser(userRepository.getReferenceById(userId));
+        List<StoredUserSongMapping> userSongMappingList =
+                userSongMappingRepository.findAllLikedSongsOfUser(userId);
 
         List<SongMetadata> songMetadataList = new ArrayList<>();
 
-        songList.forEach(song -> songMetadataList.add(SongMapper.toSongMetadata(song)));
+        userSongMappingList.forEach(userSongMapping -> {
```
Comment (SarthakSavasil) [inline src/main/java/com/emotunes/emotunes/dao/UserSongMappingDao.java]: Again

File: src/main/java/com/emotunes/emotunes/dao/UserSongMappingDao.java [hunk 5]
```diff
@@ -25,58 +26,64 @@ public class UserSongMappingDao {
...
+                    true));
+        });
+
         return songMetadataList;
     }
 
+    @Transactional
     public void songLiked(String userId, SongMetadata songMetadata, boolean isLiked) {
-        StoredSong song = songRepository.getByTitleAndDuration(songMetadata.getTitle(),
+        String songId = songRepository.getIdByTitleAndDuration(songMetadata.getTitle(),
```
Comment (SarthakSavasil) [inline src/main/java/com/emotunes/emotunes/dao/UserSongMappingDao.java]: Ispe krte discuss... I dont think ki id pass krne mai koi issue hona chaiye... Google bhikiya tha as such koi negative article nhi mila

File: src/main/java/com/emotunes/emotunes/repository/SongRepository.java [hunk 6]
```diff
@@ -19,5 +19,11 @@ public interface SongRepository extends JpaRepository<StoredSong, String> {
...
     List<StoredSong> findAllByEmotion(String emotion);
 
-    StoredSong getByTitleAndDuration(String title, LocalTime duration);
+    @Query(value =
+            "select id from song"
+                    + " where title = ?1 and"
+                    + " duration = ?2",
+            nativeQuery = true
+    )
+    String getIdByTitleAndDuration(String title, LocalTime duration);
```
Comment (SarthakSavasil) [inline src/main/java/com/emotunes/emotunes/repository/SongRepository.java]: fetching with duration is not a good idea.. krte discuss id wale ke saath

**Review by SarthakSavasil — CHANGES_REQUESTED** [review]

File: src/main/java/com/emotunes/emotunes/controller/SongPlayerController.java [hunk 7]
```diff
@@ -19,26 +21,26 @@ public class SongPlayerController {
 
     @PostMapping("/play")
     @ApiOperation("Play")
-    public void play(@RequestBody(required = false) SongMetadata songMetadata) {
+    public void play(@RequestParam("song_url") String songUrl) {
```
Comment (SarthakSavasil) [inline src/main/java/com/emotunes/emotunes/controller/SongPlayerController.java]: Do we need it? ye frontend pe hi handle hoga na

File: src/main/java/com/emotunes/emotunes/controller/SongPlayerController.java [hunk 8]
```diff
@@ -19,26 +21,26 @@ public class SongPlayerController {
...
-            @RequestBody SongMetadata songMetadata,
+            @RequestParam(value = "song_id") String songId,
             @RequestParam(value = "liked") boolean isLiked) {
-        return songPlayerService.liked(userId, songMetadata, isLiked);
+        return songPlayerService.liked(userId, songId, isLiked);
     }
 
     @PostMapping("/song_not_per_emotion")
     @ApiOperation("Song not per emotion")
     public ResponseEntity<String> songNotPerEmotion(
```
Comment (SarthakSavasil) [inline src/main/java/com/emotunes/emotunes/controller/SongPlayerController.java:34]: Function name should be more specific. This function is suggesting a different emotion for a song right? toh thoda wesa naam de de

File: src/main/java/com/emotunes/emotunes/service/impl/SongPlayerServiceImpl.java [hunk 9]
```diff
@@ -20,16 +19,16 @@ public class SongPlayerServiceImpl implements SongPlayerService {
...
-            String userId, SongMetadata songMetadata, boolean isLiked) {
-        userSongMappingDao.songLiked(userId, songMetadata, isLiked);
+            String userId, String songId, boolean isLiked) {
+        userSongMappingDao.songLiked(userId, songId, isLiked);
 
         return ResponseEntity.ok("Song Liked!");
     }
 
     @Override
     public ResponseEntity<String> songNotPerEmotion(
```
Comment (SarthakSavasil) [inline src/main/java/com/emotunes/emotunes/service/impl/SongPlayerServiceImpl.java:29]: Return string only not response entity

File: src/main/java/com/emotunes/emotunes/service/impl/SongPlayerServiceImpl.java [hunk 10]
```diff
@@ -20,16 +19,16 @@ public class SongPlayerServiceImpl implements SongPlayerService {
     @Override
     @Transactional
     public ResponseEntity<String> liked(
-            String userId, SongMetadata songMetadata, boolean isLiked) {
```
Comment (SarthakSavasil) [inline src/main/java/com/emotunes/emotunes/service/impl/SongPlayerServiceImpl.java:23]: here too

**Review by SarthakSavasil — CHANGES_REQUESTED** [review]

File: src/main/java/com/emotunes/emotunes/controller/SongPlayerController.java (hunk 8 above)
Comment (krishrustagi) [reply src/main/java/com/emotunes/emotunes/controller/SongPlayerController.java:34]: Let's not move away from the motive of this PR. Putting a todo comment.

File: src/main/java/com/emotunes/emotunes/service/impl/SongPlayerServiceImpl.java (hunk 9 above)
Comment (krishrustagi) [reply src/main/java/com/emotunes/emotunes/service/impl/SongPlayerServiceImpl.java:29]: let's change later

File: src/main/java/com/emotunes/emotunes/service/impl/SongPlayerServiceImpl.java (hunk 10 above)
Comment (krishrustagi) [reply src/main/java/com/emotunes/emotunes/service/impl/SongPlayerServiceImpl.java:23]: changed in another PR

**Review by SarthakSavasil — APPROVED** [review]

**Pull request merged**